lib/FSi/DoctrineExtensions/PropertyManipulator.php

```diff
@@ -0,0 +1,188 @@
+<?php
+
+/**
+ * (c) FSi sp. z o.o. <info@fsi.pl>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace FSi\DoctrineExtensions;
+
+use Closure;
+use InvalidArgumentException;
+use RuntimeException;
+
+/**
+ * @internal
+ */
+class PropertyManipulator
+{
+    /**
+     * An array containing closures for property read.
+     *
+     * @var Closure[]
+     */
+    private $bindings = [];
+
+    /**
+     * Internal value storage
+     *
+     * @var array
+     */
+    private $savedValues = [];
+
+    /**
+     * @param object $object
+     * @param string $name
+     * @param mixed $value
+     */
+    public function setAndSaveValue($object, $name, $value)
+    {
+        $this->setPropertyValue($object, $name, $value);
+        $this->saveValue($object, $name);
+    }
+
+    /**
+     * @param object $object
+     * @param string $name
+     * @return mixed
+     */
+    public function getPropertyValue($object, $name)
+    {
+        $this->assertIsObject($object);
+
+        $oid = spl_object_hash($object);
+        if (!isset($this->bindings[$oid][$name])) {
+            $closure = Closure::bind(function () use ($name) {
+                return $this->$name;
+            }, $object, $this->getSourceObjectForProperty($object, $name));
+
+            $this->bindings[$oid][$name] = $closure;
```

Comment:
I'm not sure if cache'ing closures per specific object per specific property speeds up anything. It would make sense to cache "getter" for specific class and property but that is against closures' nature. I would remove the cache for now. It will be 2x faster than reflection anyway (without reflection cache).